### PR TITLE
Migrate JVM specific packages out of `core-api` into new `core-api-jvm`

### PR DESCRIPTION
**Pull request by tresat**
This supports the Platformization effort by relocating JVM-specific types out of `core-api`.

### CODE REVIEW

**Comment by tresat** [comment]
@bot-gradle test this

File: subprojects/core/build.gradle.kts [hunk 1]
```diff
@@ -77,6 +77,7 @@ dependencies {
     api(libs.jsr305)
     api(libs.nativePlatform)
 
+    implementation(project(":core-api-jvm"))
```
Comment (big-guy) [inline subprojects/core/build.gradle.kts:80]: I think this is where things are tricky.  I think we don't want core to depend on the new pieces.  Is this only needed for downstream consumers of core or does something in core compile against something in core-api-jvm?

**Review by big-guy — CHANGES_REQUESTED** [review]

File: subprojects/core/build.gradle.kts (hunk 1 above)
Comment (tresat) [reply subprojects/core/build.gradle.kts]: This is needed because several classes in core: (`JavaEcosystemAttributesDescriber`, `JavaEcosystemSupport`, and `DefaultScriptClassPathResolver`) compile against `TargetJvmVersion` and `TargetJvmEnvironment`.

It seems to make sense to move those `Target` types to a JVM-specific platform.  Unfortunately moving out the types that use them (into a `core-jvm`, for instance), doesn't look like it would be so simple, as they are tied together with the script handling classes.

I think detangling the API classes is a good first step, even if that leaves `core` in pretty much the same state, just with several `core-api-x` deps instead of a single one.

I added a note to the dependency.

Comment (jvandort) [reply subprojects/core/build.gradle.kts]: I think I agree with Sterling here. I would like to see if we can solve this without needing this dependency.

**Comment by jvandort** [comment]
Have we considered `jvm-api` instead of `core-api-jvm`? What does `core` mean to us here?

**Comment by tresat** [comment]
> Have we considered `jvm-api` instead of `core-api-jvm`? What does `core` mean to us here?

I think we should keep the "core" prefix until we're further along detangling both `core-api` and `core` itself.  It makes it easier to group them for searches, etc.

**Review by big-guy — APPROVED** [review]
I think this is OK, but we should probably look at the `core` subproject first to see how we can detangle some things out of it.

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

File: subprojects/core-api/src/main/java/org/gradle/api/attributes/java/package-info.java [hunk 2]
```diff
@@ -17,5 +17,4 @@
 /**
  * Attributes specific to the Java ecosystem.
  */
-@org.gradle.api.NonNullApi
```
Comment (jvandort) [inline subprojects/core-api/src/main/java/org/gradle/api/attributes/java/package-info.java:20]: Why was this removed, especially from a public API package? This might be a breaking change for Kotlin
